### DISCUSSION TITLE
Nested Collections with relatedModel - example

### DISCUSSION THREAD

**Discussion by gumish** [jrief/django-formset#37] [General]
Hello,
at first I would like to thank you for the awesome work. The formsets in vanilla Django have been always disaster for me.

**Could you please write me some example how to use Nested Collection with related Model (for example 3 levels)?**

When I write it due to examples here ..
https://django-formset.readthedocs.io/en/latest/collections.html#nested-collection
https://django-formset.readthedocs.io/en/latest/model-collections.html#
.. it works only with the top level. But not with lower levels.
I have not found any example also here..
https://github.com/jrief/django-formset/tree/main/testapp/forms

I am not sure about the sentence:
> Since form collections can be nested, method model_to_dict can be used to recursively create a dictionary to initialize the forms, starting from a main model object. 

The model_to_dict() in the 2nd level is not even started.

```
class MemberCollection(FormCollection):
    ''' 2nd level - This level DOES NOT work - there are no initial member-forms
    '''
    min_siblings = 0
    member = MemberForm()

    def model_to_dict(self, team):
        fields = self.declared_holders['member']._meta.fields
        return [{'member': model_to_dict(member, fields=fields)}
                for member in team.members.all()]

    def construct_instance(self, team, cleaned_data):
        for data in cleaned_data:
            try:
                member = team.members.get(id=data['member']['id'])
            except (KeyError, Member.DoesNotExist):
                member = Member(team=team)
            form_class = self.declared_holders['member'].__class__
            form = form_class(data=data['member'], instance=member)
            if form.is_valid():
                if form.marked_for_removal:
                    member.delete()
                else:
                    construct_instance(form, member)
                    form.save()

class TeamCollection(FormCollection):
    """ 1st level - This level works - saved objects are rendered
    """
    min_siblings = 0
    team = TeamForm()
    members = MemberCollection()

    def model_to_dict(self, company):
        fields = self.declared_holders['team']._meta.fields
        return  [{'team': model_to_dict(team, fields=fields)}
            for team in company.teams.all()]

    def construct_instance(self, company, cleaned_data):
        for data in cleaned_data:
            try:
                team = company.teams.get(id=data['team']['id'])
            except (KeyError, Team.DoesNotExist):
                team = Team(company=company)
            form_class = self.declared_holders['team'].__class__
            form = form_class(data=data['team'], instance=team)
            if form.is_valid():
                if form.marked_for_removal:
                    team.delete()
                else:
                    construct_instance(form, team)
                    form.save()


class CompanyCollection(FormCollection):
    """ 0 level - Main edited object
    """
    form = CompanyForm()
    teams = TeamCollection()
```

Simple FormModels:
```
class CompanyForm(forms.ModelForm):
    class Meta:
        model = Company
        fields = '__all__'

class TeamForm(forms.ModelForm):
    id = forms.IntegerField(required=False, widget=forms.HiddenInput)
    class Meta:
        model = Team
        fields = ('id', 'name',)

class MemberForm(forms.ModelForm):
    id = forms.IntegerField(required=False, widget=forms.HiddenInput)
    class Meta:
        model = Member
        fields = ('id', 'name')
```

**Comment by jrief** [comment]
Thanks for these forms. I will add them to the unit tests, asap.

Please note, that yesterday I found a bug on the client side, when aggregating data on a root collection with siblings. This has been fixed in version 0.13.3.

**Reply by gumish** [reply]
I have tried the new version 0.13.3 but it doesn't solve my issue.

**Reply by gumish** [reply]
So I would very appreciate if you could write me some short working example.

**Reply by jrief** [reply]
I will use your models and create a nested form collection using them as a starting point. Please give me some time for this setup, since I'm not working full time on django-formset.

**Reply by jrief** [reply]
Just wanted to start implementing your form collections.

Could you please also paste the models `Company`, `Team` and `Member` you are using.

**Reply by gumish** [reply]
Sure. They are only models for testing so they are simple as possible.

```
class Company(models.Model):
    name = models.CharField('Company Name', max_length=50)

    class Meta:
        verbose_name = 'Company'
        verbose_name_plural = 'Companies'

    def __str__(self):
        return self.name

class Team(models.Model):
    name = models.CharField('Team Name', max_length=50)
    company = models.ForeignKey(Company, on_delete=models.CASCADE, related_name='teams')

    class Meta:
        verbose_name = 'Team'
        verbose_name_plural = 'Teams'

    def __str__(self):
        return self.name

class Member(models.Model):
    name = models.CharField('Member Name', max_length=50)
    team = models.ForeignKey(Team, on_delete=models.CASCADE, related_name='members')

    class Meta:
        verbose_name = 'Member'
        verbose_name_plural = 'Members'

    def __str__(self):
        return self.name
```

**Comment by jrief** [comment]
You have a bug in class `TeamCollection`. Method `model_to_dict` should look like:

```python
class TeamCollection(FormCollection):
    …
    def model_to_dict(self, company):
        data = []
        fields = self.declared_holders['team']._meta.fields
        for team in company.teams.all():
            data.append({
                'team': model_to_dict(team, fields=fields),
                'members': self.declared_holders['members'].model_to_dict(team),
            })
        return data
    …
```

**Reply by gumish** [reply]
Awesome - thanxks! It looks logic and simple. It hasn't been clear for me from the documentation.
Could you please also write here the "**def construct_instance**" part of the TeamCollection?
I could make up something but I would like to see the clean solution from author :)

**Reply by jrief** [reply]
I will add a special CRUD section to nested form collections.

FYI: In the next major version, I'm going to change the signature of `construct_instance`. Instead of having to pass the `cleaned_data` created by `form_collection.full_clean()` I will add a dict/list of `valid_holders` to each collection. These entities then contain the `cleaned_data` in an unnested structure.

**Comment by jrief** [comment]
@gumish I just added two extra views, they are named `company`and `companies`.
The first one can be used to edit one company and it stores level 1 `Team` and level 2 `Member`.

The latter one can be used to edit multiple companies at once.

Please note that you have to check out branch https://github.com/jrief/django-formset/tree/issue_37 in order to test this.
